homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,200 @@
+from homeassistant.helpers.discovery import async_load_platform
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_send, async_dispatcher_connect)
+
+REQUIREMENTS = ['hlk-sw16==0.0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_EVENT = 'event'
+ATTR_STATE = 'state'
+
+CONF_RECONNECT_INTERVAL = 'reconnect_interval'
+
+DATA_DEVICE_REGISTER = 'hlk_sw16_device_register'
+DEFAULT_RECONNECT_INTERVAL = 10
+CONNECTION_TIMEOUT = 10
+DEFAULT_PORT = 8080
+
+DOMAIN = 'hlk_sw16'
+
+SIGNAL_AVAILABILITY = 'hlk_sw16_device_available'
+SIGNAL_HANDLE_EVENT = 'hlk_sw16_handle_event_{}'
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+RELAY_ID = vol.All(
+    vol.Any(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 'a', 'b', 'c', 'd', 'e', 'f'),
+    vol.Coerce(str))
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        cv.string: vol.Schema({
+            vol.Required(CONF_HOST): cv.string,
+            vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+            vol.Optional(CONF_RECONNECT_INTERVAL,
+                         default=DEFAULT_RECONNECT_INTERVAL): int,
+            vol.Required(CONF_SWITCHES): vol.Schema({RELAY_ID: SWITCH_SCHEMA}),
+        }),
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the HLK-SW16 switch."""
+    # Allow platform to specify function to register new unknown devices
+    from hlk_sw16 import create_hlk_sw16_connection
+    hass.data[DATA_DEVICE_REGISTER] = {}
+
+    def add_device(device):
+        switches = config[DOMAIN][device][CONF_SWITCHES]
+
+        host = config[DOMAIN][device][CONF_HOST]
+        port = config[DOMAIN][device][CONF_PORT]
+
+        @callback
+        def disconnected():
+            """Schedule reconnect after connection has been lost."""
+            _LOGGER.warning('HLK-SW16 disconnected')
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+        @callback
+        def reconnected():
+            """Schedule reconnect after connection has been lost."""
+            _LOGGER.warning('HLK-SW16 connected')
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, True)
+
+        async def connect():
+            """Set up connection and hook it into HA for reconnect/shutdown."""
+            _LOGGER.info('Initiating HLK-SW16 connection')
+
+            client = await create_hlk_sw16_connection(
+                host=host,
+                port=port,
+                disconnect_callback=disconnected,
+                reconnect_callback=reconnected,
+                loop=hass.loop,
+                logger=_LOGGER,
+                timeout=CONNECTION_TIMEOUT,
+                reconnect_interval=DEFAULT_RECONNECT_INTERVAL)
+
+            hass.data[DATA_DEVICE_REGISTER][device] = client
+
+            # Load platforms
+            for comp_name, comp_conf in switches.items():
+                hass.async_create_task(
+                    async_load_platform(hass, 'switch', DOMAIN,
```

Comment:
It's more efficient to call load platform once with all the discovery_info and let the platform add all entities in one go.